### ISSUE TITLE
coreos instance configures incorrect routes

### ISSUE THREAD

**Issue by crawford** [coreos/bugs#554]
<a href="https://github.com/egon010"><img src="https://avatars.githubusercontent.com/u/67209?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [egon010](https://github.com/egon010)**
_Thursday Mar 27, 2014 at 23:41 GMT_
_Originally opened as https://github.com/coreos/coreos-overlay/issues/491_

----

Launching an instance in openstack results in a network route for default that points to 0.0.0.0, and no metadata route. The same configuration results in proper routes for RHEL and CentOS.

Openstack version 2013.1.4 (grizzly)
dnsmasq 2.65
Running on RHEL 6.4

"good route table"
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
169.254.169.254 10.242.180.51   255.255.255.255 UGH       0 0          0 eth0
10.242.180.0    0.0.0.0         255.255.252.0   U         0 0          0 eth0
169.254.0.0     0.0.0.0         255.255.0.0     U         0 0          0 eth0
0.0.0.0         10.242.180.1    0.0.0.0         UG        0 0          0 eth0
```

"coreos route table"
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         0.0.0.0         0.0.0.0         U         0 0          0 ens3
10.242.180.0    0.0.0.0         255.255.252.0   U         0 0          0 ens3
```

**Comment by crawford** [comment]
<a href="https://github.com/philips"><img src="https://avatars.githubusercontent.com/u/33544?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [philips](https://github.com/philips)**
_Friday Mar 28, 2014 at 00:02 GMT_

----

@egon010 Is there any way you can grab a tcpdump of the dhcp traffic? This would be a huge help.

Is dnsmasq the dhcp server?

**Comment by crawford** [comment]
<a href="https://github.com/egon010"><img src="https://avatars.githubusercontent.com/u/67209?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [egon010](https://github.com/egon010)**
_Friday Mar 28, 2014 at 00:28 GMT_

----

Yes, dnsmasq is the dhcp server. Let me get a tcpdump.

**Comment by crawford** [comment]
<a href="https://github.com/egon010"><img src="https://avatars.githubusercontent.com/u/67209?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [egon010](https://github.com/egon010)**
_Friday Mar 28, 2014 at 00:51 GMT_

----

https://www.dropbox.com/s/1rhntu01452juuw/coreos-boot.gz

**Comment by crawford** [comment]
<a href="https://github.com/egon010"><img src="https://avatars.githubusercontent.com/u/67209?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [egon010](https://github.com/egon010)**
_Tuesday Apr 08, 2014 at 18:04 GMT_

----

Anything else I can add to help?

**Comment by crawford** [comment]
<a href="https://github.com/polvi"><img src="https://avatars.githubusercontent.com/u/46035?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [polvi](https://github.com/polvi)**
_Wednesday Apr 09, 2014 at 19:53 GMT_

----

could you share the command line you used to launch dnsmasq?

/cc @brianredbeard

**Comment by crawford** [comment]
<a href="https://github.com/brianredbeard"><img src="https://avatars.githubusercontent.com/u/57542?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [brianredbeard](https://github.com/brianredbeard)**
_Wednesday Apr 09, 2014 at 20:59 GMT_

----

So from the routing table displayed it looks similar to the issue I saw on another Grizzly instance.  @egon010 try running this command on your nova host:

```
[root@openstack ~]# ps axwww | grep dnsmasq | grep -v grep
24205 ?        S      0:00 dnsmasq --no-hosts --no-resolv --strict-order --bind-interfaces --interface=tap6d1dc895-38 --except-interface=lo --pid-file=/var/lib/neutron/dhcp/dd44cbb9-75e2-4e14-b8a6-7826efb6b299/pid --dhcp-hostsfile=/var/lib/neutron/dhcp/dd44cbb9-75e2-4e14-b8a6-7826efb6b299/host --dhcp-optsfile=/var/lib/neutron/dhcp/dd44cbb9-75e2-4e14-b8a6-7826efb6b299/opts --leasefile-ro --dhcp-range=tag0,10.200.0.0,static,120s --dhcp-lease-max=256 --conf-file=/etc/neutron/dnsmasq.conf --server=8.8.4.4 --domain=openstacklocal
```

In this case the important thing to note is:  ` --dhcp-optsfile=/var/lib/neutron/dhcp/dd44cbb9-75e2-4e14-b8a6-7826efb6b299/opts`

My guess is that in that file you have an option of the form:

```
tag0,option:classless-static-route,0.0.0.0/0,10.242.180.1,169.254.169.254/32,10.242.180.51
```

What's happening in that case is that the DHCP client (ISC DHCP) on Red Hat based systems is requesting [DHCP option 121](http://tools.ietf.org/html/rfc3442) rather than relying on DHCP option 3.  Our DHCP client embedded in networkd doesn't request this option by default and thus since Networkd didn't ask, DNSmasq doesn't answer.

In digging through the DNSmasq man page i found the option ` --dhcp-option-force` which I've been curious if would solve the issue.  Further more, the pcap @egon010 attached (once again) confirms that our DHCP client is not requesting option 121.  I'll attempt to duplicate in our openstack lab, though by default this is done differently in Havana and *may* only affect Grizzly.

**Comment by crawford** [comment]
<a href="https://github.com/philips"><img src="https://avatars.githubusercontent.com/u/33544?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [philips](https://github.com/philips)**
_Thursday Jun 19, 2014 at 23:54 GMT_

----

@brianredbeard Were you able to reproduce this on your OpenStack?

/cc @eyakubovich

**Comment by crawford** [comment]
<a href="https://github.com/egon010"><img src="https://avatars.githubusercontent.com/u/67209?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [egon010](https://github.com/egon010)**
_Thursday Jun 26, 2014 at 22:46 GMT_

----

I just pulled down the alpha image, and although SSH now opens properly, it doesn't look like any configuration / config drive information is actually being used. Is there a way to get into the image to look at the routes?

**Comment by crawford** [comment]
<a href="https://github.com/marineam"><img src="https://avatars.githubusercontent.com/u/153640?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [marineam](https://github.com/marineam)**
_Friday Jun 27, 2014 at 01:32 GMT_

----

@egon010 If this is openstack with configdrive but not a metadata service there is a known bug that causes the system to hang forever waiting on the metadata service. The fix for that issue should land in alpha early next week.

**Comment by crawford** [comment]
<a href="https://github.com/eyakubovich"><img src="https://avatars.githubusercontent.com/u/3586053?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [eyakubovich](https://github.com/eyakubovich)**
_Sunday Jun 29, 2014 at 21:03 GMT_

----

Added support for setting routes from DHCP server to networkd: http://cgit.freedesktop.org/systemd/systemd/commit/?id=e1ea665edac17d75fce01b72dadfa3211b60df2c

This should make it possible to add classless-static-route option to dnsmasq like described by @brianredbeard to enable the meta-data service to be reachable via a specific gateway.

**Comment by crawford** [comment]
<a href="https://github.com/kerwin"><img src="https://avatars.githubusercontent.com/u/525532?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kerwin](https://github.com/kerwin)**
_Thursday Jul 10, 2014 at 04:44 GMT_

----

Same problem here.. I think adding support for classless-static-route may fix the issue.
Looking forward new release of CoreOS.

**Comment by crawford** [comment]
<a href="https://github.com/marineam"><img src="https://avatars.githubusercontent.com/u/153640?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [marineam](https://github.com/marineam)**
_Friday Jul 25, 2014 at 02:16 GMT_

----

This should be fixed in the current alpha, 386.1.0
